File: macos/Classes/rasp/HookDetector.swift
import Foundation
import MachO

/// Detects code injection and hooking frameworks on macOS.
///
/// Scans loaded dynamic libraries for known injection/hooking tools:
/// 1. DYLD_INSERT_LIBRARIES environment variable (primary injection vector)
/// 2. Loaded dylib scan via _dyld_image_count / _dyld_get_image_name
/// 3. Suspicious library names associated with hooking frameworks
public class HookDetector {
    public static func check() -> Bool {
        return checkDYLDEnvironment() || checkLoadedDylibs()
    }

    /// Check for DYLD injection environment variables.
    ///
    /// DYLD_INSERT_LIBRARIES is the primary mechanism for injecting
    /// code into macOS processes. Legitimate apps should never have this set.
    private static func checkDYLDEnvironment() -> Bool {
        let env = ProcessInfo.processInfo.environment

        if env["DYLD_INSERT_LIBRARIES"] != nil {
            return true
        }
        if env["DYLD_LIBRARY_PATH"] != nil {
            return true
        }
        if env["DYLD_FRAMEWORK_PATH"] != nil {
            return true
        }

        return false
    }

    /// Scan all loaded dynamic libraries for suspicious names.
    ///
    /// Hooking frameworks inject dylibs that can be detected by name.
    private static func checkLoadedDylibs() -> Bool {
        let suspiciousLibraries = [
            "substrate",
            "cycript",
            "frida",
            "fridagadget",
            "sslkillswitch",
            "sslkillswitch2",
            "mobilesubstrate",
            "substrateinserter",
            "substrateloader",
            "substratebootstrap",
            "libcycript",
            "substitute",
            "shadow",
            "liberty",
            "inject",
            "hook",
            "interpose",
            "fishhook",
            "mitmproxy",
            "charlesproxy",
        ]

        let imageCount = _dyld_image_count()
        for i in 0..<imageCount {
            if let imageName = _dyld_get_image_name(i) {
                let nameStr = String(cString: imageName).lowercased()
                for suspicious in suspiciousLibraries {
                    if nameStr.contains(suspicious) {
                        return true
                    }
                }
            }
        }

        return false
    }
}

File: linux/rasp/hook_detector.cc
#include "hook_detector.h"

#include <cstdlib>
#include <fstream>
#include <string>

namespace flutter_neo_shield {

bool HookDetector::Check() {
  return CheckLDPreload() || CheckMaps();
}

/// Check LD_PRELOAD environment variable.
///
/// LD_PRELOAD is the primary mechanism for injecting shared libraries
/// into Linux processes. Legitimate apps should not have this set.
bool HookDetector::CheckLDPreload() {
  const char* ld_preload = getenv("LD_PRELOAD");
  if (ld_preload && strlen(ld_preload) > 0) {
    return true;
  }

  // Also check /etc/ld.so.preload
  std::ifstream preload("/etc/ld.so.preload");
  if (preload.is_open()) {
    std::string line;
    while (std::getline(preload, line)) {
      // Skip empty lines and comments
      if (!line.empty() && line[0] != '#') {
        return true;
      }
    }
  }

  return false;
}

/// Scan /proc/self/maps for suspicious shared libraries.
bool HookDetector::CheckMaps() {
  std::ifstream maps("/proc/self/maps");
  if (!maps.is_open()) return false;

  const char* suspicious[] = {
    "substrate", "inject", "hook", "interpose",
    "frida", "cycript", "xposed",
  };

  std::string line;
  while (std::getline(maps, line)) {
    for (const auto& s : suspicious) {
      if (line.find(s) != std::string::npos) {
        return true;
      }
    }
  }

  return false;
}

}  // namespace flutter_neo_shield
